File: punctuation.py
import re


def fix_commas(text):
    # Replace all variations of comma punctuation with comma+space.
    for i in text:
        if i.find(",") != -1:
            comma_index = text.index(i)
            text[comma_index] = ", "
    return text


def fix_double_spaces(text):
    # Replace double spaces with single space.
    for i in text:
        if i.find("  ") != -1:
            double_space_index = text.index(i)
            text[double_space_index] = " "
    return text


def fix_leading_punctuation(text):
    # Remove leading punctuation, if there is one.
    while True:
        # Check the first character of the first element is alphabet.
        if text[0] and text[0][0].isalpha():
            break
        else:
            text.pop(0)
    return text


def fix_space_before_final_period(text):
    for i in text:
        if i.find(" .") != -1:
            space_before_final_period_index = text.index(i)
            text[space_before_final_period_index] = "."
    return text


def fix_space_before_period(text):
    for i in text:
        if i.find(" . ") != -1:
            space_before_final_period_index = text.index(i)
            text[space_before_final_period_index] = ". "
    return text


def fix_punctuation(text):
    # Create a list of the text, including punctuation.
    split_text = re.split("(\W+)", text)
    split_text = fix_commas(split_text)
    split_text = fix_double_spaces(split_text)
    split_text = fix_leading_punctuation(split_text)
    split_text = fix_space_before_period(split_text)
    split_text = fix_space_before_final_period(split_text)

    # Stitch the list back together.
    reunited_text = ""
    for x in split_text:
        reunited_text += x

    return reunited_text

File: packages/backend/src/punctuation.js
import { splitTextWithPunctuation, reuniteTextWithPunctuation } from './textManipulation.js';

export function fixCommas(arr) {
  // Replace all variations of comma punctuation with comma+space.
  for (let i = 0; i < arr.length; i++) {
    arr[i] = arr[i].replace(/\s*,\s*/g, ", ");
  }
  return arr;
}

export function fixApostrophes(arr) {
  // Replace all straight apostrophes with curved apostrophes.
  for (let i = 0; i < arr.length; i++) {
    arr[i] = arr[i].replace(/'/g, "’");
  }
  return arr;
}

export function fixMultipleSpaces(arr) {
  // Replace multiple spaces with single space.
  for (let i = 0; i < arr.length; i++) {
    arr[i] = arr[i].replace(/\s{2,}/g, " ");
  }
  return arr;
}

export function fixLeadingPunctuation(arr) {
  // Remove leading punctuation, if there is one — except opening square brackets.
  while (arr[0] && !arr[0][0].match(/[a-zA-ZÀ-ÿ]/) && !arr[0][0].match(/\[/)) {
    arr.shift();
  }
  return arr;
}

export function removeSpacesAfterFinalPeriod(arr) {
  if (arr.length > 0 && arr[arr.length - 1].includes('.')) {
    arr[arr.length - 1] = arr[arr.length - 1].replace(/\s*\.\s*$/, '.');
  }  
  return arr;
}

export function removeSpacesAfterFinalComma(arr) {
  if (arr.length > 0 && arr[arr.length - 1].includes(',')) {
    arr[arr.length - 1] = arr[arr.length - 1].replace(/\s+$/, '');
  }  
  return arr;
}

export function removeSpacesAfterFinalEllipsis(arr) {
  if (arr.length > 0) {
    if (arr[arr.length - 1].includes('…')) {
      arr[arr.length - 1] = arr[arr.length - 1].trim();
    }
  }
  return arr;
}

export function fixSpaceBeforeAndAfterPeriod(arr) {
  for (let i = 0; i < arr.length; i++) {
    arr[i] = arr[i].replace(/ *\. +/g, ". ");
    arr[i] = arr[i].replace(/ *\… +/g, "… ");
  }
  return arr;
}

export function fixSpaceBetweenTwoPeriods(arr) {
  for (let i = 0; i < arr.length; i++) {
    arr[i] = arr[i].replace(/\.\s+\./g, ".");
  }
  return arr;
}

export function fixDoublePeriods(arr) {
  for (let i = 0; i < arr.length; i++) {
    arr[i] = arr[i].replace(/\.{2}(?!\.)/g, ".");
  }
  return arr;
}

export function fixPeriodComma(arr) {
  for (let i = 0; i < arr.length; i++) {
    arr[i] = arr[i].replace(/\.,\s/g, ". ");
  }
  return arr;
}

export function fixUglyArrows(text) {
  const niceArrowedText = text.replace(/->/g, '→');
  return niceArrowedText;
}

export function fixSpacedBrackets(arr) {
  for (let i = 0; i < arr.length; i++) {
    arr[i] = arr[i].replace(/\( +/g, "(").replace(/ +\)/g, ")");
  }
  return arr;
}

export function fixPunctuation(text) {
  // Create a list of the text, including punctuation.
  var split_text = splitTextWithPunctuation(text);

  // Run the fixing train.
  split_text = fixCommas(split_text);
  split_text = fixApostrophes(split_text);
  split_text = fixMultipleSpaces(split_text);
  split_text = fixLeadingPunctuation(split_text);
  split_text = fixSpaceBeforeAndAfterPeriod(split_text);
  split_text = fixPeriodComma(split_text);
  split_text = removeSpacesAfterFinalComma(split_text);
  split_text = removeSpacesAfterFinalPeriod(split_text);
  split_text = removeSpacesAfterFinalEllipsis(split_text);
  split_text = fixSpaceBetweenTwoPeriods(split_text);
  split_text = fixSpacedBrackets(split_text);

  // Stitch the list back together.
  var reunited_text = reuniteTextWithPunctuation(split_text);

  var fixed_text = fixUglyArrows(reunited_text);

  return fixed_text;
}